manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+once. Attempting to use a continuation more than once raises
+"Continuation_already_taken" exception. For example:
+
+\begin{caml_example}{verbatim}
+try_with perform (Xchg 0)
+{ effc = fun (type a) (eff : a t) ->
+    match eff with
+    | Xchg n -> Some (fun (k: (a, _) continuation) ->
+        continue k 21 + continue k 21)
+    | _ -> None }
+\end{caml_example}
+
+The primary motivation for adding effect handlers to OCaml is to enable
+concurrent programming. One-shot continuations are sufficient for almost all of
+the concurrent programming use cases. They are also much cheaper to implement
+compared to multi-shot continuations since they do not require stack frames to
+be copied. Moreover, OCaml programs may also manipulate linear resources such
+as sockets and file descriptors. The linearity discipline is easily broken if
+the continuations are allowed to resume more than once. It would be quite hard
+to debug such linearity violations on resources due to the lack of static
+checks for linearity and the non-local nature of control flow. Hence, OCaml
+does not support multi-shot continuations.
+
+While the ``at-most once resumption'' property on continuations is ensured with
+a dynamic check, there is no check to ensure that the continuations are resumed
+``at least once''. It is left to the user to ensure that the captured
+continuations are resumed at least once. Not resuming continuations will leak
+the memory allocated for the fibers and also any resources that the suspended
+computation may hold.
+
+One may install a finaliser on the captured continuation to ensure that the
+resources are freed:
+
+\begin{caml}
+\begin{camlinput}
+exception Unwind
+Gc.finalise (fun k ->
+  try ignore (discontinue k Unwind) with _ -> ()) k
+\end{camlinput}
+\end{caml}
+
+In this case, if "k" is never resumed, then the finaliser ensures that the
+continuation stack is unwound by discontinuing with "Unwind" exception,
+allowing the computation to free up resources. However, the runtime cost of
+finalisers is much more than the cost of capturing a continuation. Hence, it is
+recommended that the user takes care of resuming the continuation exactly once
+rather than relying on the finaliser.
+
+\section{s:effects-shallow}{Shallow handlers}
+
+The examples that we have seen so far have used \textit{deep} handlers.
+Whenever a continuation is captured in a deep handler, the captured
+continuation also includes the handler. OCaml also provides \textit{shallow}
+handlers. Compared to deep handlers, a shallow handler does not include the
+captured continuation, and expects the user to provide a handler when the
+continuation is resumed.
+
+Shallow handlers make it easier to express certain kinds of programs. For
+example, suppose we want to implement state handler that permits put and get
+operations on ambient state. We can implement the state handler using deep
+handlers as follows:
```

Comment:
I don't like this example (state implemented using a deep handler whose result type is a function type) because it is difficult to understand and unnatural. (One might argue that it is common in the literature; this is true, but only experts will know it.) In OCaml, it would be much more idiomatic to use a deep handler and a (primitive) mutable cell.

Furthermore, as far as I understand, the motivation for using shallow handlers in this example is that, by exposing the recursive function `loop`, we can make the current state `s` a parameter of `loop`, thereby removing the unpleasant need to return a function (of type `S -> A * S`) in each branch of the handler. This is nice, but does not illustrate what is (in my mind) the main difference between deep handlers and shallow handlers, namely the fact that a deep handler handles *a sequence of effects*, whereas a shallow handler handles *one effect* only.

I would suggest two changes here:
* Beginning with an example other than State; ideally, an example that can be expressed using shallow handlers, but cannot easily be expressed using deep handlers. I am not sure which example to pick, though.
* Regarding the State example, I would suggest decomposing the example in two steps, as follows: (1) show the original code and explain that (although it commonly appears in introductory papers) it is not very natural; (2) desugar the deep handler into a shallow handler, without otherwise transforming the code; so `fun s` remains inside each branch; emphasize that a deep handler can always be desugared into a shallow handler inside a recursive function; (3) point out that it is now possible to hoist `fun s` outside of the branches and make it a parameter of `loop`, which leads to more natural and more efficient code, and provides one motivation for using shallow handlers.

Reply:
The point is well taken. 

> an example that can be expressed using shallow handlers, but cannot easily be expressed using deep handlers. I am not sure which example to pick, though.

I'm stuck here trying to come up with a simple example. I'll work on the state example in the mean time.

Reply:
Section 2.6 of [Handlers in Action](https://homepages.inf.ed.ac.uk/slindley/papers/handlers.pdf) has an example of shallow handlers that involves pipes. I haven't yet read it in detail, though, so I don't know if it is a convincing example of the need for shallow handlers.

Reply:
I wondered a bit about what are natural examples of shallow handlers.

- I think that the example in this tutorial, which is an instance of the general idea of handlers evolving a state/ressource across resumptions, is in fact a natural example of use of shallow handlers. I agree with @fpottier that the deep-handler presentation of state is unpalatable. Maybe a solution would be to just drop the deep-handler presentation and show the shallow-handler version only.
- @fpottier remarks that state can be implemented with a deep handler and a mutable reference, instead of passing the state as a parameter of the recursive shallow handler. My intuition is that this may not work as well as claimed: it is clearly an incorrect implementation in presence of multi-shot continuations (because the handler itself could be captured and resumed twice, leaking updates from one resumption to the other), and I find it hard to convince myself that it is correct in all situations of ({zero,one}-shot) handlers placed "above" the state handler.

In summary: I think that state as an example has more value than @fpottier suggests.

Another class of examples of shallow handlers are those where we really want to offer an effect that "can only be performed once", instead of performing it again and again with a slightly different handler. This better captures the interest of shallow handlers, but it is also much less common in practice. (The common approach is to think of effect handlers as extensions of the language enabling new operations in some dynamic scope, which is naturally a "deep" construction.) In this direction:
- One could think of a stateful protocol where the set of operations we want to handle evolves as operations are called; for example sockets where you first handle an `Open` effect, then `Listen/Accept`, etc.
- More generally, there is an embedding of session types into effect handlers, which naturally uses shallow handlers. If you think of parallel processes as "returning" by raising separate operations, then "handling" the parallel process `(P|Q)` requires handling the operations of P and the operations of Q, whereas "handling" the sequential composition `P; Q` requires handling the operations of `P`, and then handling the operations of `Q`.

Reply:
Here is an example where the handler enforces a protocol on the computation. I haven't looked at the pipes example from Handlers in Action nor the session types encoding yet. But I suspect that the ideas may be similar. 

The example is expressed naturally with a shallow handler rather than a deep handler. With a deep handler, one may have to use reference cells for tracking the session state or use an encoding similar to the state example. 

The state example is showing something similar. I wonder whether it is sufficient only to include this example and not include the deep handler encoding of the same.

```ocaml
open Effect
open Effect.Shallow

type _ Effect.t += Send : int -> unit Effect.t
                 | Recv : int Effect.t

(* Accepts [f] which performs the sequence of effects that matches the regular
   expression: [(Send;Recv)*;Send?] *)
let run (type a) (f: unit -> unit) =
  let rec loop : type a. (a,unit) continuation -> a -> [`Send | `Recv of int] -> unit =
    fun k v accept ->
      continue_with k v
      { retc = Fun.id;
        exnc = raise;
        effc = fun (type b) (eff : b t) ->
          match eff, accept with
          | Send v, `Send -> Some (fun (k: (b,_) continuation) ->
              loop k () (`Recv v))
          | Recv, `Recv v -> Some (fun (k: (b,_) continuation) ->
              loop k v `Send)
          | _ -> None }
  in
  loop (fiber f) () `Send

let ok =
  run (fun () ->
    Printf.printf "Send 42\n";
    perform (Send 42);
    Printf.printf "Recv: %d\n" (perform Recv);
    Printf.printf "Send 43\n";
    perform (Send 43);
    Printf.printf "%d\n" (perform Recv))


let fail =
  run (fun () ->
    Printf.printf "Send 0\n";
    perform (Send 0);
    Printf.printf "Send 1\n";
    perform (Send 1) (* raises Unhandled *))
```

Reply:
This is a nice example indeed! There is a similarity with state, but the interesting difference is that the set of operations being handled evolves over time, instead of evolving just the data used by the same handler. It would be very nice to use this example to motivate shallow handlers, and I don't think that you need to show a deep-handler encoding.

Reply:
I've pushed the new shallow handler section. Please have a look.